File: App/Resources/AlgorithmDetails/pdmergesort/pdmergesort.kt
fun reverseRun(arr: Array<Int>, loIn: Int, hiIn: Int) {
  var lo = loIn
  var hi = hiIn
  while (lo < hi) {
    val t = arr[lo]
    arr[lo] = arr[hi]
    arr[hi] = t
    lo++
    hi--
  }
}

// Finds the maximal run starting at indexIn (every adjacent step in the same
// direction), reversing it in place if that direction was descending.
// Returns the index where the next run starts, or -1 if this was the last
// run.
fun identifyRun(arr: Array<Int>, indexIn: Int, n: Int): Int {
  if (indexIn >= n - 1) {
    return -1
  }
  val startIndex = indexIn
  var index = indexIn
  val ascending = arr[index] <= arr[index + 1]
  index++
  while (index < n - 1) {
    val stepAscending = arr[index] <= arr[index + 1]
    if (stepAscending != ascending) {
      break
    }
    index++
  }
  if (!ascending) {
    reverseRun(arr, startIndex, index)
  }
  return if (index >= n - 1) -1 else index + 1
}

// Merges arr[start..mid) with arr[mid..end) by copying the left run into a
// scratch buffer and merging forward from the low end.
fun mergeUp(arr: Array<Int>, start: Int, mid: Int, end: Int, buffer: Array<Int>) {
  for (i in 0 until mid - start) {
    buffer[i] = arr[start + i]
  }
  var bufferPointer = 0
  var left = start
  var right = mid
  while (left < right && right < end) {
    if (buffer[bufferPointer] <= arr[right]) {
      arr[left] = buffer[bufferPointer]
      bufferPointer++
    } else {
      arr[left] = arr[right]
      right++
    }
    left++
  }
  while (left < right) {
    arr[left] = buffer[bufferPointer]
    bufferPointer++
    left++
  }
}

// Merges arr[start..mid) with arr[mid..end) by copying the right run into a
// scratch buffer and merging backward from the high end.
fun mergeDown(arr: Array<Int>, start: Int, mid: Int, end: Int, buffer: Array<Int>) {
  for (i in 0 until end - mid) {
    buffer[i] = arr[mid + i]
  }
  var bufferPointer = end - mid - 1
  var left = mid - 1
  var right = end - 1
  while (right > left && left >= start) {
    if (buffer[bufferPointer] >= arr[left]) {
      arr[right] = buffer[bufferPointer]
      bufferPointer--
    } else {
      arr[right] = arr[left]
      left--
    }
    right--
  }
  while (right > left) {
    arr[right] = buffer[bufferPointer]
    bufferPointer--
    right--
  }
}

// Picks whichever of mergeUp/mergeDown needs the smaller scratch copy.
fun mergeRuns(arr: Array<Int>, leftStart: Int, rightStart: Int, end: Int, buffer: Array<Int>) {
  if (end - rightStart < rightStart - leftStart) {
    mergeDown(arr, leftStart, rightStart, end, buffer)
  } else {
    mergeUp(arr, leftStart, rightStart, end, buffer)
  }
}

fun sort(arr: Array<Int>) {
  val n = arr.size
  if (n < 2) {
    return
  }

  var runs = mutableListOf<Int>()
  var lastRun = 0
  while (lastRun != -1) {
    runs.add(lastRun)
    lastRun = identifyRun(arr, lastRun, n)
  }

  val buffer = Array(n) { 0 }
  var runCount = runs.size
  while (runCount > 1) {
    var i = 0
    while (i < runCount - 1) {
      val end = if (i + 2 >= runCount) n else runs[i + 2]
      mergeRuns(arr, runs[i], runs[i + 1], end, buffer)
      i += 2
    }

    val compacted = mutableListOf<Int>()
    var j = 0
    while (j < runCount) {
      compacted.add(runs[j])
      j += 2
    }
    runs = compacted
    runCount = runs.size
  }
}

fun main() {
  var array = arrayOf<Int>(
    0, 39, 21, 62, 91, 77, 14, 23,
    90, 69, 51, 81, 68, 83, 32, 56,
  )
  sort(array)
  println("[%s]".format(array.joinToString(", ")))
}

File: Modules/BuiltInAlgorithms/Sources/Merge/PDMergeSort.swift
import AlgorithmKit
import SortEngineKit

/// Ported from ArrayV's `PDMergeSort` ("Pattern-Defeating Merge Sort") — a natural merge sort:
/// `identifyRun` scans for each maximal run where every adjacent pair steps the same direction
/// (all non-decreasing, or all strictly decreasing), reversing a strictly-decreasing run in place
/// to make it ascending, then `findRuns` records every run's start position. The main loop
/// repeatedly merges adjacent pairs of runs (`merge`, choosing `mergeUp`/`mergeDown` by whichever
/// side needs a smaller half-sized buffer copy) and compacts the run-boundary list, until one
/// run — the whole sorted array — remains. Already-sorted input becomes a single run and never
/// enters the merge loop at all, giving this a real `O(n)` best case unlike an ordinary merge sort.
///
/// Stable by construction: `mergeUp` copies the *left* run and takes from that copy on a tie
/// (`<=`), the standard stable tie-break. `mergeDown` builds its result backward (highest index
/// first) and takes from its *right*-run copy on a tie (`>=`) — placing the right run's element
/// into the higher slot first necessarily leaves the left run's equal element in a lower (earlier)
/// slot, which is the same stable ordering from the other direction. A strictly-decreasing run
/// never contains adjacent equal elements by definition (that would have broken the streak), so
/// reversing it can't disturb any tie.
public struct PDMergeSort: SortAlgorithm {
  public let id = AlgorithmID(rawValue: "pdmergesort")
  public let metadata = AlgorithmMetadata(
    displayName: "Pattern-Defeating Merge Sort",
    category: .merge,
    sizeRange: 16...256,
    growthModel: OperationGrowthModel(
      anchorSize: 2523, coefficients: [183052, 86.1501, 0.00296527],
      measuredSafeCeiling: nil),
    detectedGrowthModel: DetectedGrowthModel(
      family: .powerLog, coefficients: [5.8005, 1.05975], rSquared: 0.998944),
    stable: true,
    timeComplexity: ComplexityBounds(
      best: "O(n)", average: "O(n log n)", worst: "O(n log n)"),
    spaceComplexity: "O(n)",
    iconName: "arrow.triangle.merge"
  )

  public init() {}

  public func record(into engine: inout RecordingEngine) {
    let n = engine.count
    guard n >= 2 else { return }

    let copiedHandle = engine.createAuxArray(length: n / 2)
    var copied = [Int](repeating: 0, count: n / 2)

    func writeCopied(_ i: Int, _ value: Int) {
      copied[i] = value
      engine.writeAux(copiedHandle, at: i, value: value)
    }

    func mergeUp(_ start: Int, _ mid: Int, _ end: Int) {
      for i in 0..<(mid - start) {
        writeCopied(i, engine.values[i + start])
      }
      var bufferPointer = 0
      var left = start
      var right = mid
      while left < right && right < end {
        // `copied[bufferPointer]` is a real re-read of the `copiedHandle`-shadowed buffer
        // (marked via `markAuxRead`), then compared against the live `right` index via
        // `engine.compareValue` — the aux-held value plays the "held value" role.
        engine.markAuxRead(copiedHandle, at: bufferPointer)
        if engine.compareValue(right, against: copied[bufferPointer], by: (>=)) {
          engine.setValue(left, copied[bufferPointer])
          bufferPointer += 1
        } else {
          engine.setValue(left, engine.values[right])
          right += 1
        }
        left += 1
      }
      while left < right {
        engine.setValue(left, copied[bufferPointer])
        bufferPointer += 1
        left += 1
      }
    }

    func mergeDown(_ start: Int, _ mid: Int, _ end: Int) {
      for i in 0..<(end - mid) {
        writeCopied(i, engine.values[i + mid])
      }
      var bufferPointer = end - mid - 1
      var left = mid - 1
      var right = end - 1
      while right > left && left >= start {
        // Same `markAuxRead` + `engine.compareValue` pairing as `mergeUp` above.
        engine.markAuxRead(copiedHandle, at: bufferPointer)
        if engine.compareValue(left, against: copied[bufferPointer], by: (<=)) {
          engine.setValue(right, copied[bufferPointer])
          bufferPointer -= 1
        } else {
          engine.setValue(right, engine.values[left])
          left -= 1
        }
        right -= 1
      }
      while right > left {
        engine.setValue(right, copied[bufferPointer])
        bufferPointer -= 1
        right -= 1
      }
    }

    func merge(_ leftStart: Int, _ rightStart: Int, _ end: Int) {
      if end - rightStart < rightStart - leftStart {
        mergeDown(leftStart, rightStart, end)
      } else {
        mergeUp(leftStart, rightStart, end)
      }
    }

    // Finds the end of the maximal run starting at `index` (every adjacent step in the same
    // direction), reversing it in place if that direction was strictly decreasing. Returns the
    // next run's start, or -1 if this was the last one.
    func identifyRun(_ indexIn: Int, _ maxIndex: Int) -> Int {
      guard indexIn < maxIndex else { return -1 }
      let startIndex = indexIn
      var index = indexIn
      let ascending = engine.compare(index, index + 1, by: (<=))
      index += 1
      while index < maxIndex {
        let stepAscending = engine.compare(index, index + 1, by: (<=))
        if stepAscending != ascending { break }
        index += 1
      }
      if !ascending {
        engine.reversal(startIndex, index)
      }
      return index >= maxIndex ? -1 : index + 1
    }

    // ArrayV allocates `runs` with this much slack deliberately: the compaction loop below can
    // structurally read one slot past the last meaningful entry right as `runCount` drops to 1
    // (that read's value is never used again once the outer loop sees `runCount <= 1`) — harmless
    // against Java's default-zeroed, over-allocated array, but a hard crash against a Swift array
    // sized to exactly the real entry count. Pad to the same capacity so that harmless over-read
    // lands on a real (if meaningless) `Int` instead of trapping.
    let runsCapacity = (n - 1) / 2 + 2
    let runsHandle = engine.createAuxArray(length: runsCapacity)
    var runs: [Int] = []
    var lastRun = 0
    while lastRun != -1 {
      runs.append(lastRun)
      engine.writeAux(runsHandle, at: runs.count - 1, value: lastRun)
      lastRun = identifyRun(lastRun, n - 1)
    }
    let realRunCount = runs.count
    while runs.count < runsCapacity { runs.append(0) }

    var runCount = realRunCount
    while runCount > 1 {
      var i = 0
      while i < runCount - 1 {
        let end = i + 2 >= runCount ? n : runs[i + 2]
        merge(runs[i], runs[i + 1], end)
        i += 2
      }

      var wi = 1
      var wj = 2
      while wi < runCount {
        runs[wi] = runs[wj]
        engine.writeAux(runsHandle, at: wi, value: runs[wj])
        wi += 1
        wj += 2
        runCount -= 1
      }
    }

    engine.deleteAuxArray(runsHandle)
    engine.deleteAuxArray(copiedHandle)
  }
}
